### PR TITLE
Assignment2 - Omar Facundo and Manny Aguilera

### PR DESCRIPTION
**Pull request by omar5o9**
@omar5o9 and @MannyPoptart

### CODE REVIEW

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 1]
```diff
@@ -26,14 +63,65 @@ public void Error_WithData_LogsMessage()
...
+
+
+        // Assert
+        Assert.AreEqual(1, logger.LoggedMessages.Count);
+        Assert.AreEqual(LogLevel.Warning, logger.LoggedMessages[0].LogLevel);
+        Assert.AreEqual("Message 42", logger.LoggedMessages[0].Message);
+    }
+
+    [TestMethod]
+    public void Information_WithData_LogsMessage()
```
Comment (robertogarcia11) [inline Logger.Tests/BaseLoggerMixinsTests.cs:92]: Nice, I also did similar tests for this class!

File: Logger.Tests/LogFactoryTests.cs [hunk 2]
```diff
@@ -1,9 +1,39 @@
...
+
+        // Act
+        string result = logFactory.configure(validPath);
+
+        // Assert
+        Assert.AreEqual(validPath, result);
+    }
+
+    [TestMethod]
+    [ExpectedException(typeof(ArgumentNullException))]
```
Comment (robertogarcia11) [inline Logger.Tests/LogFactoryTests.cs:26]: Instead of the ExpectedException attribute in line 26, consider using Assert.Throws for  more clarification that an exception is being thrown.

File: Logger/BaseLoggerMixins.cs [hunk 3]
```diff
@@ -1,6 +1,43 @@
-﻿namespace Logger;
+#nullable enable
+using System;
+
+namespace Logger;
 
 public static class BaseLoggerMixins
 {
+    public static void Error(this BaseLogger? logger, string message, params object[] yo)
```
Comment (robertogarcia11) [inline Logger/BaseLoggerMixins.cs:8]: consider changing the object name from "yo" to something more meaningful

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 4]
```diff
@@ -1,4 +1,5 @@
-﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
+﻿#nullable enable
```
Comment (robertogarcia11) [inline Logger.Tests/BaseLoggerMixinsTests.cs:1]: there's no need to "#nullable enable" in every class since we need to enable in the PropertyGroup

**Comment by robertogarcia11** [comment]
Nice job overall, both of you!

There is an existing BaseLogger class. It needs an auto property to hold a class name. This property should be set in the LogFactory using an object initializer. ✔
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
The current date/time ✔
The name of the class that created the logger ✔
The log level ✔
The message ✔
The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
Inside of BaseLoggerMixins implement extension methods on BaseLogger for
Error, ✔
Warning, ✔
Information, and ✔
Debug. ✔ Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
Use the nameof() operator when identifying the class name to the logger ✔
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Turn on Nullability (Nullable) ✔

File: Logger.Tests/BaseLoggerMixinsTests.cs (hunk 4 above)
Comment (MannyPoptart) [reply Logger.Tests/BaseLoggerMixinsTests.cs:1]: All #nullable enable have been removed from all files, thanks for the catch.

File: Logger.Tests/LogFactoryTests.cs (hunk 2 above)
Comment (omar5o9) [reply Logger.Tests/LogFactoryTests.cs]: Ok thanks I used Assert.Throws now

File: Logger/BaseLoggerMixins.cs (hunk 3 above)
Comment (omar5o9) [reply Logger/BaseLoggerMixins.cs]: Thanks, did this

File: Logger.Tests/BaseLoggerMixinsTests.cs (hunk 1 above)
Comment (omar5o9) [reply Logger.Tests/BaseLoggerMixinsTests.cs:92]: Cool 😎

**Comment by StrangeRanger** [comment]
@robertogarcia11 Good job and thank you for adding comments to the configuration-based requirements.

> Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔❌ (missing it in logger.test.csproj)
> Ensure that you turn on code analysis (EnableNETAnalyzers) ✔❌ (missing it in logger.test.csproj)
> Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔❌ (missing it in logger.test.csproj)
> Ensure that you turn on EnforceCodeStyleInBuild ✔❌ (missing it in logger.test.csproj)
> Turn on Nullability (Nullable) ✔❌ (missing it in logger.test.csproj)

File: Logger.Tests/BaseLoggerMixinsTests.cs (hunk 4 above)
Comment (BenjaminMichaelis) [inline Logger.Tests/BaseLoggerMixinsTests.cs:1]: Why is this needed here?

File: Logger.Tests/Logger.Tests.csproj [hunk 5]
```diff
@@ -1,9 +1,15 @@
...
+		<TargetFramework>net8.0</TargetFramework>
 		<IsPackable>false</IsPackable>
-		<LangVersion>11.0</LangVersion>
+		<LangVersion>12.0</LangVersion>
+    <Nullable>enable</Nullable>
+    <EnableNETAnalyzers>True</EnableNETAnalyzers>
+    <Nullability>True</Nullability>
+    <CodeAnalysisTreatWarningAsErrors>True</CodeAnalysisTreatWarningAsErrors>
+    <EnforceCodeStyleInBuild>True</EnforceCodeStyleInBuild>
+    <TreatWarningsAsErrors>True</TreatWarningsAsErrors>
```
Comment (BenjaminMichaelis) [inline Logger.Tests/Logger.Tests.csproj:12]:
```suggestion
		<LangVersion>12.0</LangVersion>
		<Nullable>enable</Nullable>
		<EnableNETAnalyzers>true</EnableNETAnalyzers>
		<Nullability>true</Nullability>
		<CodeAnalysisTreatWarningAsErrors>true</CodeAnalysisTreatWarningAsErrors>
		<EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
		<TreatWarningsAsErrors>true</TreatWarningsAsErrors>
```
casing was wrong. And spacing should be consistent.

File: Logger/BaseLogger.cs [hunk 6]
```diff
@@ -1,7 +1,16 @@
...
+namespace Logger;
 
 public abstract class BaseLogger
 {
     public abstract void Log(LogLevel logLevel, string message);
+
+    public string? ClassName
+    {
+        get; set;
+    }
```
Comment (BenjaminMichaelis) [inline Logger/BaseLogger.cs:11]: This should be a single line, and also why is it nullable?

File: Logger/BaseLoggerMixins.cs [hunk 7]
```diff
@@ -1,6 +1,43 @@
-﻿namespace Logger;
+
+using System;
+
+namespace Logger;
 
 public static class BaseLoggerMixins
 {
+    public static void Error(this BaseLogger? logger, string message, params object[] args)
```
Comment (BenjaminMichaelis) [inline Logger/BaseLoggerMixins.cs:8]: Why is logger nullable? I expect to call this message, a logger is needed.

File: Logger/BaseLoggerMixins.cs [hunk 8]
```diff
@@ -1,6 +1,43 @@
...
+        get; set;
+    }
 
+    public static void LogDo(this BaseLogger? logger, LogLevel level, string message, params object[] args)
+    {
+        if (logger == null)
+        {
+            throw new ArgumentNullException(nameof(logger), "Logger can't be null");
+        }
+        string output = string.Format(CulInfo, message, args);
```
Comment (BenjaminMichaelis) [inline Logger/BaseLoggerMixins.cs:39]: InvarientCulture would of been ok in this case

File: Logger/FileLogger.cs [hunk 9]
```diff
@@ -0,0 +1,28 @@
+﻿
+using System;
+using System.Globalization;
+using System.IO;
+
+namespace Logger;
+
+public class FileLogger(string? fileName) : BaseLogger
+{
+    private readonly string? FilePath = fileName;
```
Comment (BenjaminMichaelis) [inline Logger/FileLogger.cs:10]: Why isn't this a property?

File: Logger/LogFactory.cs [hunk 10]
```diff
@@ -1,10 +1,45 @@
...
+
+using System;
+
+namespace Logger;
+
 
 public class LogFactory
 {
-    public BaseLogger CreateLogger(string className)
+    private string? _logPath;
```
Comment (BenjaminMichaelis) [inline Logger/LogFactory.cs:9]: Why isn't this a property

File: Logger/LogFactory.cs [hunk 11]
```diff
@@ -1,10 +1,45 @@
...
+    public string ConfigureFileLogger(string logPath)
     {
+        if (logPath == null)
+        {
+            throw new ArgumentNullException(nameof(logPath), " File path can not be null"); ;
+        }
+        else
+        {
+            this._logPath = logPath;
```
Comment (BenjaminMichaelis) [inline Logger/LogFactory.cs:18]: `this` is not needed

File: Logger/LogFactory.cs [hunk 12]
```diff
@@ -1,10 +1,45 @@
...
+    public BaseLogger? CreateLogger(string className)
+    {
+        if (_logPath == null)
+        {
+            throw new ArgumentNullException(nameof(className), " File path can not be null"); ;
+        }
+        else
+        {
+            
+            FileLogger logg = new (ConfigureFileLogger(_logPath!))
```
Comment (BenjaminMichaelis) [inline Logger/LogFactory.cs:34]: Is this a spelling mistake?
```suggestion
            FileLogger log = new (ConfigureFileLogger(_logPath!))
```

File: Logger/Logger.csproj [hunk 13]
```diff
@@ -1,8 +1,17 @@
...
-    <LangVersion>11.0</LangVersion>
+    <TargetFramework>net8.0</TargetFramework>
+    <LangVersion>12.0</LangVersion>
+    <Nullable>enable</Nullable>
+    <EnableNETAnalyzers>True</EnableNETAnalyzers>
+    <Nullability>True</Nullability>
+    <CodeAnalysisTreatWarningAsErrors>True</CodeAnalysisTreatWarningAsErrors>
+    <EnforceCodeStyleInBuild>True</EnforceCodeStyleInBuild>
+    <TreatWarningsAsErrors>True</TreatWarningsAsErrors>
+
```
Comment (BenjaminMichaelis) [inline Logger/Logger.csproj:12]:
```suggestion
    <EnableNETAnalyzers>true</EnableNETAnalyzers>
    <Nullability>true</Nullability>
    <CodeAnalysisTreatWarningAsErrors>true</CodeAnalysisTreatWarningAsErrors>
    <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
```
casing is incorrect
